nemo/collections/nlp/models/nlp_model.py

```diff
@@ -0,0 +1,89 @@
+        # after initializing DDP with PTL.
+        if app_state.model_parallel_size is not None:
+            LightningModule.init_ddp_connection(self, global_rank, world_size, is_slurm_managing_tasks)
+            if app_state.model_parallel_group is None:
+                mpu.initialize_model_parallel(app_state.model_parallel_size)
+                app_state.model_parallel_group = mpu.get_model_parallel_group()
+                app_state.data_parallel_group = mpu.get_data_parallel_group()
+                app_state.model_parallel_rank = torch.distributed.get_rank(group=app_state.model_parallel_group)
+                app_state.data_parallel_rank = torch.distributed.get_rank(group=app_state.data_parallel_group)
+                logging.info(f'mp_rank: {app_state.model_parallel_rank}')
+                logging.info(f'dp_rank: {app_state.data_parallel_rank}')
+
+        else:
+            return LightningModule.init_ddp_connection(self, global_rank, world_size, is_slurm_managing_tasks)
+
+    def configure_ddp(self, model, device_ids):
+        """ Override LightningModule ddp if using model parallel. """
+
+        app_state = AppState()
+
+        if app_state.model_parallel_size is not None:
+            logging.info("Configuring DDP for model parallelism.")
+            logging.info(f"data_parallel_group: {app_state.data_parallel_group}")
+            # with model parallelism, multiple GPUs form a large "logical GPU"
+            # this means that data parallel groups span multiple GPUs
+            # and are non-trivial
+
+            model = LightningDistributedDataParallel(
+                model, device_ids, output_device=device_ids[0], process_group=app_state.data_parallel_group
+            )
+            return model
+
+        else:
+            logging.info("Did not detect model parallel using LightningModule.configure_ddp")
+            return LightningModule.configure_ddp(self, model, device_ids)
+
+    def setup(self, stage):
+        """ PTL hook that is called after DDP is initialized """
+
+        if stage == 'fit':
+
+            app_state = AppState()
+
+            if app_state.model_parallel_size is not None:
+                logging.info("replacing sampler with model parallel sampler")
+                mp_sampler = torch.utils.data.distributed.DistributedSampler(
+                    self._train_dl.dataset,
+                    num_replicas=app_state.model_parallel_size,
+                    rank=app_state.data_parallel_rank,
+                )
+                mp_dl = self._trainer.replace_sampler(self._train_dl, mp_sampler)
+                self._train_dl = mp_dl
+
+                logging.info(f"restoring model parallel checkpoint: {self.bert_model._restore_path}")
+                # model parallel checkpoints need to be restored after torch.distributed is initialized
+                self.bert_model.restore_weights(self.bert_model._restore_path)
```

Comment:
This assumes that NLPModel class will have to have bert_model. 
In general we should not assume that for a class with a generic name such as NLPModel.
Instead of doing this:
```
self.bert_model.restore_weights(self.bert_model._restore_path)
```

Can we do something more generic like
```
self.restore_weights(...)
```

And have ```restore_weights``` properly implemented by every child model which needs model parallelism?

Reply:
I was trying to avoid having every child model have to implement something. I wanted to make it so that model parallelism would be automatic. The idea is that when using model parallelism, a directory is passed in for the BERT checkpoint, rather than a file. We can then parse that directory and the number of sub-directories is the model parallel size.

So I'm thinking, we add bert_model flag to NLPModel as @ekmb suggested. Then we can check if bert_model is actually a MegatronBertEncoder. If so, then we call MegatronBertEncoder restore.